user/api.go

```diff
@@ -691,6 +753,163 @@ func (a *Api) DeleteUser(res http.ResponseWriter, req *http.Request, vars map[st
+	}
+
+	log.Infof("Will merge account %v with idNat %v", user.Id, user.FrProId)
+
+	if originalUser, err := a.Store.FindUser(req.Context(), &User{Id: user.Id}); err != nil {
+		a.sendError(res, http.StatusInternalServerError, STATUS_ERR_FINDING_USR, log, err)
+
+	} else if originalUser == nil {
+		a.sendError(res, http.StatusUnauthorized, STATUS_UNAUTHORIZED, log, "User not found")
+
+	} else if dupUsers, err := a.Store.FindUsers(req.Context(), &User{FrProId: user.FrProId}); err != nil {
+		a.sendError(res, http.StatusInternalServerError, STATUS_ERR_FINDING_USR, log, err)
+
+	} else if len(dupUsers) == 1 && dupUsers[0].Id != user.Id {
+		//only throw an error if there is a user with a different id but with the same oidc external id
+		a.sendError(res, http.StatusConflict, STATUS_ID_ALREADY_USED, log)
+
+	} else if len(dupUsers) > 1 {
+		a.sendError(res, http.StatusConflict, STATUS_ID_ALREADY_USED, log)
+
+	} else {
+		// Everything is fine, update the user
+		updatedUser := originalUser.DeepClone()
+		updatedUser.FrProId = user.FrProId
+		if err := a.Store.UpsertUser(req.Context(), updatedUser); err != nil {
+			a.sendError(res, http.StatusInternalServerError, STATUS_ERR_UPDATING_USR, log, err)
+		} else {
+			a.sendUserWithStatus(res, updatedUser, http.StatusAccepted, false)
+		}
+	}
+}
+
+// OIDC callback method
+func (a *Api) DelegatedLoginCallback(res http.ResponseWriter, req *http.Request) {
+	rp.CodeExchangeHandler(a.processDelegatedLogin, a.provider)(res, req)
+}
+
+func (a *Api) redirectToBlipError(res http.ResponseWriter, errorMsg string) {
+	a.logger.Print(errorMsg)
+	res.Header().Set("location", a.ApiConfig.FrontUrl+"/professional/certify?source=psc&error="+errorMsg)
+	res.WriteHeader(http.StatusFound)
+}
+
+// Callback method used to process response from OIDC provider
+// Will redirect to blip /merge page if the user does not have the oidc subject id already set.
+// Will redirect to blip root / if the user is found
+// Will redirect to blip /error page if something goes wrong
+// Always return HTTP 302
+func (a *Api) processDelegatedLogin(res http.ResponseWriter, req *http.Request, tokens *oidc.Tokens, state string, rp rp.RelyingParty) {
+	// Extract user info from token
+	jwtToken, err := jwt.Parse(tokens.IDToken, nil)
+	if jwtToken == nil {
+		a.logger.Error("Error while parsing JWT token received from OIDC provider.", err)
+		a.redirectToBlipError(res, "Internal server error")
+		return
+	}
+	user := &User{}
+	claims := jwtToken.Claims.(jwt.MapClaims)
+	user.FrProId = claims["sub"].(string)
+
+	// Prepare oidc cookie
+	oidcTokens := OidcTokens{
+		AuthToken:    tokens.AccessToken,
+		RefreshToken: tokens.RefreshToken,
+	}
+	if cookieVal, err := oidcTokens.Encode(); err != nil {
+		a.logger.Error("Error while encoding JWT token before sending to Blip.", err)
+		a.redirectToBlipError(res, "Internal server error")
+		return
+	} else {
+		rp.CookieHandler().SetCookie(res, "ecps-oidc", cookieVal)
+	}
```

Comment:
Maybe we could make this a function, these long 'if else' is difficult to understand